scalameta/contrib/shared/src/main/scala/scala/meta/contrib/ScaladocToken.scala

```diff
@@ -0,0 +1,197 @@
+package scala.meta.contrib
+
+/** Represents a scaladoc element */
+sealed abstract class ScaladocToken
+
+/**
+ * The available tokens and their documentation are obtained from:
+ * @see http://docs.scala-lang.org/overviews/scaladoc/for-library-authors.html
+ */
+object ScaladocToken {
+
+  /** A blank line */
+  final case object Break extends ScaladocToken
+
+  /** A paragraph which failed parsing */
+  final case class Unknown(text: String) extends ScaladocToken
+
+  // Text
+
+  /** An inline part of a text block */
+  trait TextPart {
+    def syntax: String
+  }
+
+  /** An untagged description */
+  final case class Text(words: TextPart*) extends ScaladocToken
+
+  /** A single word */
+  final case class Word(text: String) extends TextPart {
+    override def syntax: String = text
+  }
+
+  /** A single embedded code expression */
+  final case class CodeExpr(code: String) extends TextPart {
+    override def syntax: String = s"{{{$code}}}"
+  }
+
+  /** A block of one or more lines of code */
+  final case class CodeBlock(code: Seq[String]) extends ScaladocToken
+
+  /** A heading */
+  final case class Heading(level: Int, title: String) extends ScaladocToken
+
+  /* List blocks */
+
+  /** Represents a list item */
+  final case class ListItem(desc: Text, nested: Option[ListBlock] = None) extends ScaladocToken
+
+  /** Represents a list block */
+  final case class ListBlock(prefix: String, items: ListItem*) extends ScaladocToken
+
+  /* Tags */
+
+  // https://docs.scala-lang.org/overviews/scaladoc/for-library-authors.html#tags
+  sealed abstract class TagType(
+      val tag: String,
+      val hasLabel: Boolean = false,
+      val hasDesc: Boolean = false
+  )
+
+  /**
+   * Represents a tagged documentation remark
+   * @param label set iff [[tag.hasLabel]]
+   * @param desc set iff [[tag.hasDesc]]
+   */
+  final case class Tag(tag: TagType, label: Word = null, desc: Text = null) extends ScaladocToken
```

Comment:
tags are implemented using a single type, referring to a tag type as a field, rather than as separate types, and this part resembles the existing approach. it was easier to instantiate them like that when parsing, let me know if you prefer to have separate types and have a simple way to construct.